packages/jsii-pacmak/lib/targets/java.ts

```diff
@@ -2529,11 +2529,12 @@ class JavaGenerator extends Generator {
       'throw new ClassNotFoundException("Unknown JSII type: " + fqn);',
     );
     this.code.closeBlock();
-    this.code.line(
-      'return this.cache.computeIfAbsent(MODULE_TYPES.get(fqn), this::findClass);',
-    );
+    this.code.line('String className = MODULE_TYPES.get(fqn);');
+    this.code.openBlock('if (!this.cache.containsKey(className))');
+    this.code.line('this.cache.put(className, this.findClass(className));');
```

Comment:
Any chance this can cause an infinite recursion (which I assume is why `computeIfAbsent` throws this error)?

Reply:
I think that there is a circuit breaker [here](https://github.com/aws/jsii/blob/master/packages/@jsii/java-runtime/project/src/main/java/software/amazon/jsii/JsiiEngine.java#L224), but I'll @RomainMuller comment

Reply:
I think this is going to be enough indeed. The problem was that further mutation to the map happened as part of `this.findClass` within the lambda expression passed to `computeIfAbsent`. This approach makes "atomic" mutations here (and this is safe because `jsii` apps can be expected to be single-threaded).